src/System.Memory/src/System/Buffers/ReadOnlySequence.cs

```diff
@@ -330,6 +330,17 @@ private ReadOnlySequence<T> SliceImpl(in SequencePosition begin, in SequencePosi
             );
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private SequenceType GetSequenceType()
+        {
+            // We take high order bits of two indexes index and move them
+            // to a first and second position to convert to BufferType
+            return _sequenceStart.GetObject() == null ? SequenceType.Empty : (SequenceType)((((uint)_sequenceStart.GetInteger() & ReadOnlySequence.FlagBitMask) >> 30) | (uint)_sequenceEnd.GetInteger() >> 31);
```

Comment:
This will be faster:
```C#
return _sequenceStart.GetObject() == null ?
                SequenceType.Empty :
                (SequenceType)(-(2 * (_sequenceStart.GetInteger() >> 31) + (_sequenceEnd.GetInteger() >> 31)));
```

![image](https://user-images.githubusercontent.com/6527137/38407269-ac5ab278-392d-11e8-8971-57a93d7ce53b.png)

Reply:
`(((uint)_sequenceStart.GetInteger() & ReadOnlySequence.FlagBitMask) >> 30) | (uint)_sequenceEnd.GetInteger() >> 31`
and
`(-(2 * (_sequenceStart.GetInteger() >> 31) + (_sequenceEnd.GetInteger() >> 31))`
are equivalent for all [start, end] between int.MinValue and int.MaxValue.

**Here is the breakdown:**
![image](https://user-images.githubusercontent.com/6527137/38408754-be4d5d68-3933-11e8-90c2-63d19d2bf180.png)

Reply:
It is works https://dotnetfiddle.net/kJjenR